apps/shipments/serializers.py

```diff
@@ -294,7 +312,7 @@ def validate(self, attrs):
                 # Validate authenticity and integrity of message signature
                 attrs['payload'] = json.loads(jws.verify(payload, public_key, header['alg']).decode("utf-8"))
             else:
-                raise exceptions.PermissionDenied(f"Certificate {header['kid']} is "
+                raise exceptions.PermissionDenied(f"Certificate {certificate_id_from_payload} is "
```

Comment:
Here. If the certificate status != 'ACTIVE', then pull Device.get_valid_certificate and update device.certificate_id. That should simplify some of this logic I think.

Reply:
Good point Adam, however doing so won't take in account the edge case where the data are coming from another device with a valid certificate unless we re implement here same `if certificate_id_from_payload != device.certificate_id:` then fetch new one and update